### DISCUSSION TITLE
chapter_tree/binary_tree/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#153] [Announcements]
# chapter_tree/binary_tree/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_tree/binary_tree/

**Comment by ZZhenF** [comment]
```java
/* 插入与删除结点 */
TreeNode P = new TreeNode(0);
// 在 n1 -> n2 中间插入结点 P
n1.left = P;
P.left = n2;
// 删除结点 P
n1.left = n2;
```
插入P这里，先令n1的左值针指向p结点，会将n2以后的结点丢失吧，虽然这里创建二叉树的时候就已经将每个结点的地址都保存下来了，但是先令p的左指针先指向n2，在让n1的左值针指向p会不会更严谨一点

**Reply by krahets** [reply]
应该都行哈，已经暂存在 n2 了，顺序不太重要，感觉这里不用太纠结～
如果想要避免暂存，尽量缩短代码，可以这样写

```java
P.left = n1.left
n1.left = P
```

**Reply by wbzyo** [reply]
确实会丢失，这里能直接用n2，是因为上文初始化的时候已经定义了n2，先令p的左指针先指向n2（P.left=n1.left),再让n1的左值针指向p会更严谨一点(n1.left=P)

**Comment by ruqingxu001** [comment]
”二叉树中的插入与删除一般都是由一套操作配合完成的“，这里的”一套操作“指什么呢？可以理解为资源的子节点的资源释放吗？

**Reply by krahets** [reply]
哈喽，这里的“一套操作”指的是多个步骤的操作，并且一种操作可能分为多种情况来处理。比如在「二叉搜索树」中，删除结点操作要分为三种情况处理，每种情况需要不同的多个步骤的操作。看完二叉搜索树你应该就理解了～

**Comment by xillianpeng** [comment]
我对高度的定义有点疑问，如果高度的定义是走过边的数量，那么只有根节点的二叉树，高度是否就是0，完美二叉树总结点数量=2^0-1=0

**Reply by xillianpeng** [reply]
不好意思，忽略吧，公式看错了，总结点数量应该是2^(0+1) - 1 = 1

**Comment by Swordknife** [comment]
那么有一个结点的二叉树，高度和深度都是0，对吗

**Reply by krahets** [reply]
是的，树的高度为 0 ，该结点的高度、深度都为 0

**Comment by xusujun33** [comment]
7.1.3 tips：
在中文社区中，完美二叉树常被称为「满二叉树」，请注意与完满二叉树区分。这里最后一句话是不是写错了，是否是请注意与完全二叉树区分。

**Reply by xusujun33** [reply]
是我看错了，冒失了冒失了。一个是完美，一个是完满。。我的天呐，太容易看错了。

**Comment by tianjx98** [comment]
请问一下画图是用的什么工具啊, 很好看!

**Reply by gvenusleo** [reply]
应该是 Keynote 🤔

**Comment by mhkz** [comment]
类比楼的高度，和水的深度，从下往上看和从上往下看

**Reply by krahets** [reply]
确实，从根节点到看叶节点（从顶看底）是深度，从叶节点看根节点（从底看顶）是高度

**Comment by youngledo** [comment]
感觉二叉树的内容有点少，案例缺失，能补充点实际场景吗？

**Reply by krahets** [reply]
本书主要介绍数据结构的概念与用法，看完建议可以开始去刷二叉树的题啦

**Comment by Jaynxe** [comment]
在完美二叉树的定义中，这里写的是(除了最底层外，其余所有层的节点都被完全填满)
最底层的节点数应该也是满的吧

**Reply by krahets** [reply]
emmm 这里的表述确实有歧义，本来想表达的是子节点被填满，即除了最底层外，其余层的所有节点的度都为 2 。我优化一下表述

**Comment by Sail38** [comment]
`删除节点通常意味着删除该节点及其所有子树` 这句话看不懂， 以图7-3第一幅图（原图， 在插入P node之前）为例，如果 Node 2 含有左（和/或）右子树，那么整个子树在Node2删除操作中都没了吗？  请帮忙解答， 谢谢。

**Reply by krahets** [reply]
Hi ，如果我们仅删除 `node2` （断开它和父节点之间的连接），那么 `node2` 及其以下形成的子树就都被删除了。我们无法再从树的根节点访问到这个子树上的任意一个节点。

**Comment by ChinaYong** [comment]
完美二叉树中，说了“若树高度为h，则节点总数为2^(h+1)”，这里不应该用“高度”，而应该用“深度”吧？包括后面的陈述都是用的“高度”这个词。

**Reply by krahets** [reply]
我们通常习惯说“二叉树的高度”。我感觉这是因为自然界中的树是从地面往天上长的，所以人类对于“树”的描述都是“好高啊”。

**Reply by ChinaYong** [reply]
就是说上面用的并非定义，而是一种常见的表述方式是吧？

**Reply by wyang0824** [reply]
我是这么理解的：计算总节点个数，用树高度来描述更合理。深度说的是某个节点在整个树中所属的位置

**Comment by zoujia** [comment]
作者你好，根据文章里“7.1.1二叉树常见术语”中的定义，以及使用“边”来衡量“高度”与“深度”，有一个疑惑：叶子节点的“高度”是 0 吗？类似的，只有一个根节点的二叉树，该树的“高度”是 0 吗？

我问的这个问题，直接关系到我对于“图7-7平衡二叉树”里的“变量 d ”的计算的理解，因为这个地方我没看懂，我自己根据上文的定义来计算每个节点的 d 的值，发现跟文中对不上~~

**Reply by krahets** [reply]
你说的没错。这个在[本章小结](https://www.hello-algo.com/chapter_tree/summary/#2-q-a)里也有讲。

**Comment by lavaicer** [comment]
节点的「高度 height」：从最远叶节点到该节点所经过的边的数量。

**Reply by lavaicer** [reply]
您好，这里6的高度从8算还是从12算。一个普通节点的高度应该是从该节点的最远叶子节点算起还是从整个树最远的叶子节点算起。谢谢

**Reply by krahets** [reply]
Hi，节点高度的准确定义是：从距离该节点最远的叶节点到该节点所经过的边的数量。

**Comment by EririIcytail** [comment]
我大学教材和百度的大部分答案说， 树的高度是从根节点到最远节点的节点个数而不是边的个数，因此树的高度  =  树的深度  = 树中节点的最大高度 + 1 ，！= 节点的最大高度。本文定义是否有误？最后的表格中，是否“高度为 h 的树的节点总数”应为“ 2 ^ h - 1” ? 不知道是不是有多种定义， 供作者参考喵

**Reply by krahets** [reply]
Hi，我们以 Wiki 为准：“树的高度是从根节点到最远节点的边的个数” ，请参考 [
Tree (data structure)](https://en.wikipedia.org/wiki/Tree_(data_structure))

> The height of a node is the length of the longest downward path to a leaf from that node. The height of the root is the height of the tree. The depth of a node is the length of the path to its root (i.e., its root path). Thus the root node has depth zero, leaf nodes have height zero, and a tree with only a single node (hence both a root and leaf) has depth and height zero. Conventionally, an empty tree (tree with no nodes, if such are allowed) has height −1.

这个概念的不一致性主要是由于对”path“这个概念定义不同导致的，细节请参考[这里](https://github.com/krahets/hello-algo/discussions/304#discussioncomment-5580117)

**Reply by 3238068224** [reply]
6，我是说怎么按照大神写的去做学校的题，完全不对。

**Reply by 3238068224** [reply]
我还到处搜，晕了。大神可以特别标注一下这个地方。大学的教材好像默认二叉树根节点的高度和深度都是1.至少我现在学的教材就是这样说的，这还是学校专门选的考研数据结构。服了

**Reply by krahets** [reply]
@3238068224 看来这不一致还挺广泛的……考试的话还是以课本为准～

**Reply by wyang0824** [reply]
能否将  树的高度 定义放到 7.1.1节 常见术语里？我看后面用到了这个定义，但是常见术语中 没找到

**Comment by tanhaow** [comment]
才知道原来汉语里面parent node是父节点...😅

**Comment by jcluo1994** [comment]
self.left: TreeNode | None = None  # 左子节点引用
这个不应该是self.left=None吗？还是说规范就是这样？

**Reply by krahets** [reply]
冒号后面的部分是 Python 的类型注解（type hints），详情请见 [typing — Support for type hints](https://docs.python.org/3/library/typing.html) 。
为了提升规范性和可读性，这本书的 Python 代码全部添加了 type hints 。

**Comment by chenghui-ch** [comment]
图7-5 完全二叉树 第三层 第三个节点是最底层节点吗？为啥没被填满

**Reply by krahets** [reply]
Hi～在最底层，填满需要 8 个节点，图中只有 5 个节点；右边还有三个位置没有填满。

第三层已经被填满了，第三层的第三个节点的“子节点”没有被填满，请注意区分～

**Reply by weidongweidong** [reply]
最底层不是第四层么？

**Comment by luoflyxidian** [comment]
想请问下，如果在嵌入式系统中使用高度为20的满二叉树，由于内存有限，无法存储所有节点，该如何操作和遍历这样规模的二叉树，采用什么方法可以减少内存的占用

**Reply by krahets** [reply]
满二叉树可以用数组存储，相比链表更加节省空间。可以看下后面的“二叉树的数组表示”

**Reply by ganxingrui** [reply]
前提是内存有限，就得考虑懒加载；例如先加载左子树，砍掉右子树，直接节省一半内存；然后左子树，还是用同样的方法处理；左子树处理完，移除内存，再去加载又子树；思想就是懒加载；局限性就是空间节省了，相对而言浪费时间；此时就需要考虑二者直接寻找一个平衡。

**Comment by xiewanqin** [comment]
你好，想问下子树的高度该怎么定义呢，这里有点迷惑，子树是根节点的左（右）节点及其以下节点形成的树的话，那高度是不是应该是左（右）节点到最远叶节点所经过的边的数量。但是以平衡二叉树的定义来看好像不是这样

**Reply by krahets** [reply]
树的高度是“它的根节点”到最远叶节点所经过边的数量。子树也有自己的根节点

**Comment by ganxingrui** [comment]
大神，二叉树插入代码，这样写更好点吧
```java
// 1.先让P  n1同时指向n2,防止n2丢失
P.left = n1.left
// 2.让n1指向P
n1.left = P
```

**Reply by CodeGeis** [reply]
你将评论时间顺序显示,最早一条评论就是你这个问题

**Comment by qw124341789** [comment]
你好，作者，如果能把每章评论里的问题集做到每章的结尾，这样Hello算法就更完善了。

**Reply by krahets** [reply]
Hi，谢谢建议！现在每章的小结里已经有精选的 Q&A 了哈，我们会定期更新

**Comment by ggplz** [comment]
貌似 除了根节点的其他节点都叫叶节点，而度为 0 的节点应该叫叶子节点？

**Reply by krahets** [reply]
Hi，在本书中，叶节点和叶子节点等价（是同一个名词）

**Comment by bearoops** [comment]
二叉树的「高度 height」：从根节点到最远叶节点所经过的边的数量。
错误修正：“根节点”改成“该节点”。

**Reply by bearoops** [reply]
可以删除评论不？

**Comment by ferryboy001** [comment]
很高兴可以发现这个宝藏网站。
我看不见，只能借助[屏幕阅读器](https://zh.wikipedia.org/wiki/%E8%9E%A2%E5%B9%95%E9%96%B1%E8%AE%80%E5%99%A8)，以及键盘来浏览网站。
网站使用的公式渲染方式非常无障碍，是我在其他 PDF、epub 或者纯文本上极少见到的。
这里因为无法看到图，所以有一些对于二叉树概念上的疑问想请各位大佬解惑：

1. 完美二叉树和完满二叉树具体的区别是什么？或者说，他们之间的包含关系是什么？是什么样的条件让某个树成为完满二叉树而不是完美二叉树（反之亦然）？

2. 完全二叉树中，
> 只有最底层的节点未被填满，且最底层节点尽量靠左填充

具体是什么样的结构？尽量靠左填充是如何操作的？怎样才算不尽量？如何的不尽量靠左填充会让这个结构不是一个完全二叉树？

最后，很高兴可以有这样一个评论区，可以在啃书的时候提出疑问。
谢谢！

**Reply by suzumiy** [reply]
您好,我很想帮助您理解但又怕我的表述有误,所以我的回答仅供参考
第一个问题:完美二叉树和完满二叉树具体的区别是什么？或者说，他们之间的包含关系是什么？是什么样的条件让某个树成为完满二叉树而不是完美二叉树（反之亦然）
回答如下:完美二叉树的条件更为苛刻,是完美二叉树就一定是完满二叉树了,包含关系是完满二叉树包含完美二叉树,完美二叉树是完满二叉树中的苛刻条件.具体区别是这样的,完满二叉树要求除了叶子节点外,其他所有节点都有两个子节点可以理解为要么没有子节点(自己就是子节点了)要么就有两个子节点,即不存在只有一个子节点的节点,简单举个完满二叉树的例子 根节点A下有两个节点B左和B右,B左下有两个子节点C左和C右,C左和C右为子节点下面没有节点了,B右也为子节点下面也没有子节点,这样除了子节点外所有节点都有两个子节点。完美二叉树要求所有层都被填满,上述的完满二叉树有三层ABC,但是B右没有子节点所以造成第三层没有填满,如果B右也有两个子节点则该完满二叉树就是完美二叉树了。
同样的第二个问题:完全二叉树中，只有最底层的节点未被填满，且最底层节点尽量靠左填充,具体是什么样的结构？尽量靠左填充是如何操作的？怎样才算不尽量？如何的不尽量靠左填充会让这个结构不是一个完全二叉树？
回答如下,上面举例的完满二叉树就是一个完全二叉树,因为只有第三层没有被填满而且是第二层靠右的B右没有子节点,如果B左的两个节点C左和C右移动到B右下,作为B右的子节点,这颗完满二叉树就不是完全二叉树了,因为左边的第二层节点B左没有子节点,不满足尽量靠左的要求。值得注意的是完全二叉树不要求所有父节点都拥有两个子节点,上述举例的完满二叉树中如果B左的子节点只有C左也同样是完全二叉树,但如果只有C右这一个子节点就不是完全二叉树,因为C右并没有尽量靠左。
希望我的回答能帮助到您。

**Reply by ferryboy001** [reply]
谢谢，我好像理解了。
如果给定一个三层的完美二叉树，每层节点用 `a`、`b`、`c` 表示：

1. 如果移除 b 右的子节点，这个结构就是完满二叉树，也是完全二叉树；

2. 在此基础上，移除 b 左的右子节点，这个结构依然是完全二叉树，但不是完满二叉树；

3. 还是在第一步的基础上，如果移除的是 b 左的左子节点，这个结构既不是完全二叉树，也不是完满二叉树；

特殊的，一个链表应该也是一个完全二叉树吧~。

**Reply by suzumiy** [reply]
是的,您的描述很有条理比我的要清晰,以我的拙见您的描述是对的,但是一个链表可能不是完全二叉树,完全二叉树的定义为「完全二叉树 complete binary tree」只有最底层的节点未被填满，且最底层节点尽量靠左填充。所以如果这个链表只有2个元素的话是完全二叉树,如果大于两个元素会出现非底层节点未被填满的情况。

**Reply by ferryboy001** [reply]
好的，忘了有这个条件了……

**Reply by krahets** [reply]
谢谢鼓励！@suzumiy 的回答很详细了，我再简短回答一下：

1. 在完美二叉树中，只有最底层的节点的度为 0 ，或者说完美二叉树每一层的节点都是被填满的。相对地，完满二叉树只要求节点的度为 0 或 2 ，因此每一层的节点都可能存在度为 0 的情况。
2. 我用一个例子来描述完全二叉树“底层节点尽量靠左”的情况：给定一个共 3 层节点的二叉树，那么最底层最多可以容纳 4 个节点，将它们从左到右编号为 0, 1, 2, 3 。此时假设最底层有 2 个节点，那么这两个节点必须在 0, 1 位置时才是完全二叉树；其他所有的位置组合都不满足“尽量靠左”。尽量靠左就是从左到右无空位的填满。

新年快乐！给你点赞！我在设计内容的呈现时，可能考虑的并不周到，欢迎随时提建议意见

**Reply by ferryboy001** [reply]
谢谢回复！
因为这是一本开源的书，是随时可以修改的，很多在其他书本里对于像我这样的读者形成阅读障碍的部分，也许我们一直没办法搞懂，就很遗憾。
虽然一图胜千言，但这千言对于我们来说非常重要，当然，也可以不用千言那么多，只要能把图片中的信息完整的用文字传达出来，就足够了。
希望这本书能越来越好！

**Comment by deusyu** [comment]
k神这个可以添加导出为PDF的功能吗，我想把某个章节喂到ChatGPT进行对话式学习。

**Reply by krahets** [reply]
Hi，浏览器打印页面（Ctrl + P）应该可以导出 PDF ，或者可以直接下载 PDF 版 https://github.com/krahets/hello-algo/releases

**Reply by deusyu** [reply]
打印页面，包含了评论。
刚刚忘记补充了，导出为不带评论的页面。
我的想法是学一节自己对话式学习，并且让ChatGPT 出题巩固一节。
从下载的PDF版喂给他链路可能过长也可能上下文过多。
以上仅仅是自己的需求和建议😇

**Reply by krahets** [reply]
好的～在 Chrome 浏览器中，可以选中正文，鼠标右键 + 打印，这样就不会有评论的页面了。或者在打印时，仅打印前几页

**Comment by wxamlzm** [comment]
我把二叉树当成倒着的树来看是不是就可以了TAT

**Reply by krahets** [reply]
没问题哈哈：

- 二叉树最顶部的节点被称为“根节点”，对应的就是现实中的“树根” ；
- 二叉树最底部的节点被称为“叶节点”，对应现实树上的“叶子”；

**Reply by LeoooKang** [reply]
是的，对着图看概念的时候，发现很奇怪。在这里把图片里的树反着画，效果应该会更好的。（个人理解）

**Reply by krahets** [reply]
@WalkerInAI 在计算机学科中，树通常就是文中这样画的，需要习惯它

**Reply by YMMdrq** [reply]
是倒过来的树，因为根最高的。

**Comment by mrCR100** [comment]
请问在二叉树常用术语中最远叶节点是指什么含义呢？有严格的定义吗？

**Reply by krahets** [reply]
在二叉树中，根节点到叶节点之间的边的数量可以看作是根节点到叶节点的“距离”。每个叶节点到根节点之间的边的数量可能不同，而边数量最多的叶节点，就是“最远叶节点”。

**Reply by mrCR100** [reply]
理解了。

**Comment by l1n9d** [comment]
请问插入一个节点的时候，不用比较相邻节点的大小吗？

**Reply by Autumn00air** [reply]
看树的要求，比如二叉搜索树，就需要比左子树大，比右子树小，所以插入的时候需要进行下沉算法。当然，树本身就是宏观的，所以在没有要求之前，你想怎么操作都可以（个人理解，欢迎指正~）

**Reply by shaoxinH** [reply]
主要根据需求设计呀，不同的二叉树插入顺序适合不同的需求

**Comment by redandblueqwer** [comment]
K神，判断节点高度中，"从距离该节点最远的叶节点到该节点所经过的边的数量"，其中的叶节点是默认对应子树下的情况吗？

**Reply by DBook111** [reply]
我认为不是子树中最远的叶节点，而是所有叶节点中距离最远的

**Reply by whereishero** [reply]
叶节点就是默认对应子树下的，因为树结构是单向的，只能从一个结点到它的子节点

**Comment by genskyff** [comment]
把节点度数改成出度和入度比较好，出度和入度之和为该节点的度数，二叉树就是一个根节点入度为0，叶节点出度为0的无回路连通图。这样还可以利用握手定理，进一步推出二叉树的边数=节点数-1

**Reply by loskyertt** [reply]
哪需要什么握手定则，e = n - 1对任何一颗树都成立，不仅限于二叉树。一棵树中除了根节点不提供边，其它节点都会提供一条边，不就直接推出来了？

**Comment by jack1415926** [comment]
你好，我想问一下在二叉树节点结构体中，本页展示的代码只有left和right两个节点但在文件中那个的代码却有parent共3个节点，有和书中一样的示例代码吗

**Comment by sqhsysu17** [comment]
大佬您好！初学算法，有个疑问。图7-3中，插入树节点时，代码顺序是n1.left = p 然后p.left = n2，可是这样不会导致n2找不到吗？就和链表的插入操作一样，不应该先让p.next连接上n2，再让n1.next链接p吗？不知道我的理解是否有误

**Reply by LittleCabbage-liu** [reply]
这是因为他这段代码里面每个节点都是单独定义出来了的，而不是只提供了一个根节点

**Reply by jiuerya7** [reply]
原来如此👍

**Comment by OrangeAlien** [comment]
我不明白，C的代码里面有节点高度`height`这一成员，后面插入和删除节点时，为什么不对它进行赋值

**Comment by Mike-7777777** [comment]
新增内容:

> 节点所在的层（level）：从顶至底递增，根节点所在层为 1 。

此处的1应注明该定义不统一, 一说为0.

https://stackoverflow.com/questions/59151282/what-is-level-of-root-node-in-a-tree

**Comment by xuaidongx** [comment]
h树高的perfect binary tree的节点计算是等比公式求和{1,2,4,8...} 
2<sup>n</sup>+1 = 2<sup>h+1</sup>+1 
n:项数 h:根节点到叶子结点的edge数

**Comment by dibe1202** [comment]
有人能帮忙解释一下，图 7-7平衡二叉树，第二层左边的节点的d=1是怎么算的， 它只有一个左子节点，没有右子节点。  那它的左子树高度为0？ 没有右子树？

**Reply by DaHuaJia** [reply]
？你看错了吧，很好理解啊，它有一个左叶子节点，高度是1，没有右节点。d=1-0=1

**Reply by dibe1202** [reply]
不对啊，按上面的定义，从距离该节点最远的叶节点到该节点所经过的边的数量。 左子树只有一个节点，它没有边，它的高度就是0，然后没有右子树。 这是我的点。

**Reply by HarrryHe** [reply]
我们将d=1的那个节点取名为d1，d1的高度为1，由于节点高度的定义为从距离该节点最远的叶节点到该节点所经过的边的数量 （也就是他的左子节点 同时也是树的叶节点），且d1没有右子节点，根据公式来 d的绝对值 = 1 - 0 = 1。应该不难理解，或许你在回去把定义读一遍，然后再把平衡二叉树的定义也读一遍就懂了 加油!

**Reply by qwerty123-debug** [reply]
@DaHuaJia @HarrryHe 平衡二叉树的定义是任意节点的左子树与右子树的高度差的绝对值不超过1，一个二叉树的高度的定义是从根节点到最远叶节点所经过的边的数量。根据这两个定义，d1的左子树的根节点和最远叶节点是同一个，所以从根节点到最远叶节点没有经过任何边，那么高度就是0，也就是d1的左子树高度为0。而之所以你们算成了1，是因为你们弄错了定义，这个左子结点高度确实为1，但它的左子树高度为0，要区分树的高度和节点的高度的定义是不同的。你们这样做是相当于把d1的左子树的高度当作了d1的左子结点在起初那个大二叉树中的高度。
然后继续说d1的右子树，右子树没有根节点和最远叶节点，如果依然按照定义的话，那么经过的边数仍为0，那么d1的右子树的高度也为0。原来那位@dibe1202说的是对的。
但是后来我注意到了这样一句话（在图7-2下面那条Tip里）：请注意，我们通常将“高度”和“深度”定义为“经过的边的数量”，但有些题目或教材可能会将其定义为“经过的节点的数量”。在这种情况下，高度和深度都需要加 1 。
用教材里的定义可以解决这个问题。根据教材里的定义，d1的左子树从根节点到最远叶结点经过了一个节点，也就是根节点，所以高度是1，而右子树因为没有节点，所以没有经过节点，那么高度就是0，这样的话就既满足了定义，也解释了为什么d1的左子树与右子树的高度差为1。

**Reply by Mike-7777777** [reply]
@qwerty123-debug 在这个情况下，空树的高度应该定义为`-1`，而不是`0`。见 [链接](https://www.geeksforgeeks.org/height-and-depth-of-a-node-in-a-binary-tree/)。这个问题在stackflow上被讨论过，一部分教材将空树的高度定义为undefined，见链[接](https://stackoverflow.com/questions/2209777/what-is-the-definition-for-the-height-of-a-tree)，但我认为应该定义为-1。

**Reply by HarrryHe** [reply]
今天复习 回来看了看，感觉当时还是看的太快了 （哭 确实是搞错了 当时理所当然就说出他本身的高度了

**Reply by HarrryHe** [reply]
在7.5中有提到叶节点高度为0，空节点高度为-1，适用于这里平衡二叉树的情况

**Reply by xiaoyaozizaiwoweidu** [reply]
让我先回顾一下什么是高度。树的高度是指从根节点到最远叶子节点的最长路径上的边数。因此，一个节点的高度就是从该节点到其最远叶子节点的最长路径上的边数。
在图中，每个节点都有一个 d 值，表示其左子树高度减去右子树高度。例如，根节点的 d=−1，这意味着它的右子树比左子树高1。根节点的左子树高度2，右子树高度3，所以2-3=-1，这符合平衡条件，因为 ∣−1∣=1。

**Reply by Dumbliidore** [reply]
我看到此处也有同样的疑问，要满足图中所示，只有空树的高度为-1，但我觉得这样较为难以理解。个人比较赞成@qwerty123-debug所说，以教材中树的高度的定义来理解更加符合，在此基础上，我进一步认为文章中对树的深度和高度等定义，应该与教材对深度和高度的定义保持一致，会更加准确一些。

**Comment by swcjmy666** [comment]
打开

**Comment by superZchen0701** [comment]
'''完全二叉树（complete binary tree）只有最底层的节点未被填满，且最底层节点尽量靠左填充。'''
如果按照上面的说法，完美二叉树并不是一棵完全二叉树。因为完美二叉树的叶子节点也是填满的。
但是如果按照'''最底层节点靠左填充，其他层节点全被填满'''说法来，完美二叉树就是一棵完全二叉树。

**Reply by laohugit** [reply]
完全二叉树"仅允许"最底层的节点不完全填满，"仅允许"不是你说的"只有"哦。显然完美二叉树及是完全二叉树也是完满二叉树更是平衡二叉树哦，要不说是完美呢。

**Comment by xiaozhuge123** [comment]
day06

**Comment by WangYun0918** [comment]
请问：是否节点所在的层（level） = 节点的深度（depth）+ 1，该公式是否成立？

**Comment by WangYun0918** [comment]
请问“在完美二叉树中，叶节点的度为 0，其余所有节点的度都为  2；” 不是所有的二叉树的叶节点的度都为0吗，如果节点的度不为0，肯定不是二叉树啊

**Reply by laohugit** [reply]
节点的度表示子节点的个数，取值0，1，2。

**Reply by WangYun0918** [reply]
谢谢回复。但我的意思是叶节点，如果度不是0，那就不是叶节点了，所以叶节点的度为0感觉是一句“废话”

**Comment by WangYun0918** [comment]
最后一个表格“ 二叉树的最佳结构与最差结构”，第二行“高度为 h 的树的叶节点数量”，链表为何是 1 ，不应该是 h 吗？

**Reply by WangYun0918** [reply]
而且完美二叉树也不应该是 2h 应该是 2h - 1 才对吧

**Reply by WangYun0918** [reply]
抱歉 我理解错了，懂了

**Comment by LukeLuke-Y** [comment]
二叉树的高度（height）：从根节点到最远叶节点所经过的边的数量。
这里是不是写成：从最远叶节点到根结点所经过的边的数量。会好一点，本质上数量一样，但是这样更能统一计算方向，高度是从叶节点向目标节点算，深度是从根结点向目标节点算。

**Reply by fug-0117** [reply]
是的 深度从上往下 高度从下往上 这样好记忆

**Comment by nanmingjushi** [comment]
删除节点时，应彻底断开被删除节点与前后节点的所有关联，避免残留引用。 
```java

// 删除节点 P
n1.left = n2;
P.left = null;
```

**Comment by interset-wq** [comment]
二叉树的插图是如何制作出来的

**Comment by xinfeifeifeifei** [comment]
我感觉把二叉树 倒过来，根节点在下面  。这样好理解

**Reply by limo06-sketch** [reply]
666

**Comment by ruqingxu001** [comment]
你的邮件我已经收到！

**Comment by beidimanman** [comment]
谢谢，您的邮件我已收到。
